### PR TITLE
Allow anomalies to be computed using the climatology from a subset of the selected period

### PR DESCRIPTION
**Pull request by jvegasbsc**
I was working on the preprocessor for a new diagnostic and realized that I will need the capability to compute the anomalies relative to a specific period and not the full one I am using for the analysis

I added that option to the anomalies preprocessor. By default, the behaviour will be the same: use the full available period

**Tasks**

-   [ ] [Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do, if you haven't done so already (and add the link at the bottom)
-   [x] This pull request has a descriptive title that can be used in a changelog
-   [x] Add unit tests
-   [x] Public functions should have a numpy-style docstring so they appear properly in the [API documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/api/esmvalcore.html). For all other functions a one line docstring is sufficient.
-   [x] If writing a new/modified preprocessor function, please update the [documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/esmvalcore/preprocessor.html)
-   [x] Circle/CI tests pass. Status can be seen below your pull request. If the tests are failing, click the link to find out why.
-   [x] Codacy code quality checks pass. Status can be seen below your pull request. If there is an error, click the link to find out why. If you suspect Codacy may be wrong, please ask by commenting.
-   [x] Please use `yamllint` to check that your YAML files do not contain mistakes
-   [x] If you make backward incompatible changes to the recipe format, make a new pull request in the [ESMValTool repository](https://github.com/ESMValGroup/ESMValTool) and add the link below

If you need help with any of the tasks above, please do not hesitate to ask by commenting in the issue or pull request.

* * *

### CODE REVIEW

**Comment by jvegasbsc** [comment]
I have to update the extract time because I found a bug when trying to get the February 30th in a 360 day calendar

File: esmvalcore/preprocessor/_time.py [hunk 1]
```diff
@@ -78,21 +80,20 @@ def extract_time(cube, start_year, start_month, start_day, end_year, end_month,
...
-        time=lambda t: t_1 <= time_units.date2num(t.point) < t_2)
+        time=lambda t: t_1 <= t.point < t_2)
 
     cube_slice = cube.extract(constraint)
     if cube_slice is None:
         start_cube = str(cube.coord('time').points[0])
         end_cube = str(cube.coord('time').points[-1])
         raise ValueError(
-            f"Time slice {start_date} to {end_date} is outside cube "
+            f"Time slice {start_cube} to {end_cube} is outside cube "
```
Comment (bouweandela) [inline esmvalcore/preprocessor/_time.py]:
```suggestion
            f"Time slice {t_1} to {t_2} is outside cube "
```

File: esmvalcore/preprocessor/_time.py [hunk 2]
```diff
@@ -409,13 +410,17 @@ def climate_statistics(cube, operator='mean', period='full'):
...
+        iris.util.promote_aux_coord_to_dim_coord(clim_cube, clim_coord.name())
+    else:
+        clim_cube = iris.cube.CubeList(
+            clim_cube.slices_over(clim_coord.name())).merge_cube()
+    cube.remove_coord(clim_coord)
     return clim_cube
 
 
-def anomalies(cube, period):
+def anomalies(cube, period, reference):
```
Comment (bouweandela) [inline esmvalcore/preprocessor/_time.py]:
```suggestion
def anomalies(cube, period, reference=None):
```

**Comment by jvegasbsc** [comment]
> Could you please add documentation here? https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/esmvalcore/preprocessor.html#anomalies

I wrote it but then I forgot to add it to the commit 🤦

File: doc/esmvalcore/preprocessor.rst [hunk 3]
```diff
@@ -872,6 +872,10 @@ Parameters:
       Available periods: 'full', 'season', 'seasonal', 'monthly', 'month',
       'mon', 'daily', 'day'. Default is 'full'
 
+    * reference: Time slice to use as the reference to compute the climatology
+      on. Can be 'null' to use the full cube or a dictionary with the
+      parameters from extract_time_. Default is null
```
Comment (bouweandela) [inline doc/esmvalcore/preprocessor.rst]: It would be nice if you could add an example of how to use this below, because this is not entirely trivial.

Comment (jvegasbsc) [reply doc/esmvalcore/preprocessor.rst]: Done

**Comment by bouweandela** [comment]
Thanks for adding the documentation, could you please solve the merge conflicts?

**Comment by bouweandela** [comment]
Could you have a look at the Codacy issue please? Apart from that I think this looks good.

File: doc/esmvalcore/preprocessor.rst [hunk 4]
```diff
@@ -893,25 +893,47 @@ Parameters:
...
+            anomalies:
+              period: month
+              reference:
+                start_year: 1979
+                start_month: 1
+                start_day: 1
+                end_year: 2000
+                end_month: 12
+                end_day: 31
+                standardize: true
```
Comment (mattiarighi) [inline doc/esmvalcore/preprocessor.rst]:
```suggestion
              standardize: true
```

**Comment by mattiarighi** [comment]
Tested successfully, but please fix the indentation in the example.

**Pull request merged**